runtime/compiler/env/J9SharedCache.cpp

```diff
@@ -1441,69 +1442,132 @@ TR_J9JITServerSharedCache::rememberClass(J9Class *clazz, const AOTCacheClassChai
+      else if (TR_SharedCache::INVALID_CLASS_CHAIN_OFFSET != clientClassChainOffset)
          {
          OMR::CriticalSection classChainDataMapMonitor(clientData->getClassChainDataMapMonitor());
-         cache.insert({ clazz, { classChainOffset, NULL } });
+         cache.insert({ clazz, { clientClassChainOffset, NULL } });
          }
       }
 
-   return classChainOffset;
+   if (useServerOffsets)
+      return record ? record->data().idAndType() : TR_SharedCache::INVALID_CLASS_CHAIN_OFFSET;
+   else
+      return clientClassChainOffset;
    }
 
 J9SharedClassCacheDescriptor *
 TR_J9JITServerSharedCache::getCacheDescriptorList()
    {
+   TR_ASSERT(_stream, "stream must be initialized by now");
+   TR::Compilation *comp = TR::compInfoPT->getCompilation();
+   ClientSessionData *clientData = comp->getClientData();
+   bool useServerOffsets = clientData->useServerOffsets(_stream) && comp->isAOTCacheStore();
+   TR_ASSERT_FATAL(!useServerOffsets, "Unsupported when ignoring the client SCC");
+
    auto *vmInfo = TR::compInfoPT->getClientData()->getOrCacheVMInfo(_stream);
    return vmInfo->_j9SharedClassCacheDescriptorList;
    }
 
+bool
+TR_J9JITServerSharedCache::isClassInSharedCache(TR_OpaqueClassBlock *clazz, uintptr_t *cacheOffset)
+   {
+   TR_ASSERT(_stream, "stream must be initialized by now");
+   TR::Compilation *comp = TR::compInfoPT->getCompilation();
+   ClientSessionData *clientData = comp->getClientData();
+   bool useServerOffsets = clientData->useServerOffsets(_stream) && comp->isAOTCacheStore();
+
+   if (useServerOffsets)
+      {
+      bool missingLoaderInfo = false;
+      auto classRecord = clientData->getClassRecord(reinterpret_cast<J9Class *>(clazz), _stream, missingLoaderInfo);
+      if (classRecord)
+         {
+         if (cacheOffset)
+            *cacheOffset = classRecord->data().idAndType();
+         return true;
+         }
+      return false;
+      }
+   else
+      {
+      J9ROMClass *romClass = reinterpret_cast<J9ROMClass *>(fe()->getPersistentClassPointerFromClassPointer(clazz));
+      return isROMClassInSharedCache(romClass, cacheOffset);
+      }
+   }
+
 bool
 TR_J9JITServerSharedCache::isMethodInSharedCache(TR_OpaqueMethodBlock *method, TR_OpaqueClassBlock *definingClass, uintptr_t *cacheOffset)
    {
-   // TODO: The definingClass parameter of this method is currently unused, but it will become necessary when AOT cache stores are
-   // allowed at the server without a client SCC.
-   J9ROMMethod *romMethod = fe()->getROMMethodFromRAMMethod(reinterpret_cast<J9Method *>(method));
-   // We use isROMStructureInSharedCache directly because we want to forbid
-   // isROMMethodInSharedCache from being used at the server
-   return isROMStructureInSharedCache(romMethod, cacheOffset);
+   TR::Compilation *comp = TR::compInfoPT->getCompilation();
+   ClientSessionData *clientData = comp->getClientData();
+   bool useServerOffsets = clientData->useServerOffsets(_stream) && comp->isAOTCacheStore();
```

Comment:
This sequence repeats in several of the functions of `TR_J9JITServerSharedCache`. I wonder if we can keep some info in the `TR_J9JITServerSharedCache` object so that we can avoid finding it again and again. The most expensive part is the thread local TR::compInfoPT, so maybe we can store that.
In `processEntry()`, when a new compilation starts we do:
```
      if (_vm->sharedCache())
         // Set/update stream pointer in shared cache.
         // Note that if remote-AOT is enabled, even regular J9_SERVER_VM will have a shared cache
         // This behaviour is consistent with non-JITServer
         ((TR_J9JITServerSharedCache *) _vm->sharedCache())->setStream(stream);
```
Maybe at this point we can attach back `this`.

Reply:
I also thought about adding a `useServerOffsets()` property to `TR::Compilation` directly, which would be set if the compilation is an AOT cache store and we are ignoring the local SCC. That might simplify some of the logic elsewhere as well.

Reply:
Never mind, it's the fact that we look up the compilation in the compilation thread info that's the issue.

Reply:
Can the `CompilationInfoPerThread` be stored in the `TR_J9JITServerSharedCache` like that? I assumed the server SCC object was shared between compilation threads.

Reply:
The way I remember it is that each thread has an associated frontEnd which is created on demand if it doesn't already exist. During the constructor of the frontEnd we also create a sharedCache which is attached to the frontEnd. This each thread (and implicitly each compilation thread) will have its own sharedCache object.

Reply:
Yes, you're right. In `TR_J9VMBase::get()` the compilation info per thread is consulted, and those all have individual frontend objects in them.

Reply:
I added another commit to address this.